internal/wrappers/codebashing-http.go

```diff
@@ -0,0 +1,115 @@
+package wrappers
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	commonParams "github.com/checkmarx/ast-cli/internal/params"
+	"github.com/golang-jwt/jwt"
+	"github.com/pkg/errors"
+	"github.com/spf13/viper"
+)
+
+const (
+	failedToParseCodeBashing    = "Failed to parse list results"
+	failedGettingCodeBashingURL = "Authentication failed, not able to retrieve codebashing base link"
+	tenThousand                 = "10000"
+	limit                       = "limit"
+	incorrectFlags              = "No codebashing link available"
+)
+
+type CodeBashingHTTPWrapper struct {
+	path string
+}
+
+func NewCodeBashingHTTPWrapper(path string) *CodeBashingHTTPWrapper {
+	return &CodeBashingHTTPWrapper{
+		path: path,
+	}
+}
+
+func (r *CodeBashingHTTPWrapper) GetCodeBashingLinks(params map[string]string, codeBashingURL *string) (
+	*[]CodeBashingCollection,
+	*WebError,
+	error,
+) {
+	clientTimeout := viper.GetUint(commonParams.ClientTimeoutKey)
+	params[limit] = tenThousand
+	resp, err := SendHTTPRequestWithQueryParams(http.MethodGet, r.path, params, nil, clientTimeout)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer resp.Body.Close()
+	decoder := json.NewDecoder(resp.Body)
+	switch resp.StatusCode {
+	case http.StatusBadRequest, http.StatusInternalServerError:
+		errorModel := WebError{}
+		err = decoder.Decode(&errorModel)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, failedToParseCodeBashing)
+		}
+		return nil, &errorModel, nil
+	case http.StatusOK:
+		var decoded []CodeBashingCollection
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, failedToParseCodeBashing)
+		}
+		err = json.Unmarshal(body, &decoded)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, failedToParseCodeBashing)
+		}
+		if decoded[0].Path == "" {
+			return nil, nil, errors.Errorf(incorrectFlags)
+		}
+		decoded[0].Path = *codeBashingURL + decoded[0].Path
+		return &decoded, nil, nil
+	default:
+		return nil, nil, errors.Errorf("response status code %d", resp.StatusCode)
+	}
+}
+
+func (r *CodeBashingHTTPWrapper) GetCodeBashingURL(field string) (*string, error) {
+	tokenExpirySeconds := viper.GetInt(commonParams.TokenExpirySecondsKey)
+	accessToken := getClientCredentialsFromCache(tokenExpirySeconds)
+	if accessToken == nil {
+		authURI, err := getAuthURI()
+		if err != nil {
+			return nil, err
+		}
+		accessKeyID := viper.GetString(commonParams.AccessKeyIDConfigKey)
+		accessKeySecret := viper.GetString(commonParams.AccessKeySecretConfigKey)
+		astAPIKey := viper.GetString(commonParams.AstAPIKey)
+		if accessKeyID == "" && astAPIKey == "" {
+			return nil, errors.Errorf(fmt.Sprintf(FailedToAuth, "access key ID"))
+		}
+		accessToken, err = getClientCredentials(accessKeyID, accessKeySecret, astAPIKey, authURI)
+		if err != nil {
+			return nil, errors.Errorf(failedGettingCodeBashingURL)
+		}
+	}
```

Comment:
We are duplicating logic here from client.go, consider the following:

- public function in client.go that checks api-key/client-id and returns the accessToken
- enrichWithOath2Credentials calls that method and injects the result into the request
- we call the new function directly and parse the token if no err

This would centralize the parameter checking and the calls, would also remove the need for this to be a method and it could be placed as a function in result.go

Reply:
Done, thanks !